service/worker/deletenamespace/reclaimresources/workflow.go

```diff
@@ -0,0 +1,171 @@
+		RetryPolicy:        retryPolicy,
+		WorkflowRunTimeout: 60 * time.Minute,
+	}
+)
+
+func validateParams(params *ReclaimResourcesParams) error {
+	if params.NamespaceID.IsEmpty() {
+		return temporal.NewNonRetryableApplicationError("namespace ID is empty", "", nil)
+	}
+
+	if params.Namespace.IsEmpty() {
+		return temporal.NewNonRetryableApplicationError("namespace is empty", "", nil)
+	}
+
+	return nil
+}
+
+func ReclaimResourcesWorkflow(ctx workflow.Context, params ReclaimResourcesParams) (ReclaimResourcesResult, error) {
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Workflow started.", tag.WorkflowType(WorkflowName))
+
+	if err := validateParams(&params); err != nil {
+		return ReclaimResourcesResult{}, err
+	}
+
+	var a *Activities
+
+	// Step 1. Delete workflow executions.
+	result, err := deleteWorkflowExecutions(ctx, params)
+	if err != nil {
+		return result, err
+	}
+
+	// Step 2. Delete namespace.
+	ctx5 := workflow.WithLocalActivityOptions(ctx, localActivityOptions)
+	err = workflow.ExecuteLocalActivity(ctx5, a.DeleteNamespaceActivity, params.Namespace, params.NamespaceID).Get(ctx, nil)
+	if err != nil {
+		return result, fmt.Errorf("%w: DeleteNamespaceActivity: %v", errors.ErrUnableToExecuteActivity, err)
+	}
+
+	logger.Info("Workflow finished successfully.", tag.WorkflowType(WorkflowName))
+	return result, nil
+}
+
+func deleteWorkflowExecutions(ctx workflow.Context, params ReclaimResourcesParams) (ReclaimResourcesResult, error) {
+	var a *Activities
+	logger := workflow.GetLogger(ctx)
+	deleteAttempt := int32(1)
+	executionsExist := true
+	var result ReclaimResourcesResult
+
+	for executionsExist {
+		ctx1 := workflow.WithChildOptions(ctx, deleteExecutionsWorkflowOptions)
+		var der deleteexecutions.DeleteExecutionsResult
+		err := workflow.ExecuteChildWorkflow(ctx1, deleteexecutions.DeleteExecutionsWorkflow, params.DeleteExecutionsParams).Get(ctx, &der)
+
+		if err != nil {
+			logger.Error("Unable to execute child workflow.", tag.WorkflowType(deleteexecutions.WorkflowName), tag.Error(err))
+			return result, fmt.Errorf("%w: %s: %v", errors.ErrUnableToExecuteChildWorkflow, deleteexecutions.WorkflowName, err)
+		}
+		result.SuccessCount += der.SuccessCount
+		result.ErrorCount += der.ErrorCount
+
+		for checkAttempt := int32(1); checkAttempt <= checkMaxAttempts; checkAttempt++ {
+			err = workflow.Sleep(ctx, visibilityDelay(der.SuccessCount))
+			if err != nil {
+				return result, err
+			}
+
+			ctx2 := workflow.WithLocalActivityOptions(ctx, localActivityOptions)
+			err = workflow.ExecuteLocalActivity(ctx2, a.CheckExecutionsExistActivity, params.NamespaceID, params.Namespace).Get(ctx, &executionsExist)
+			if err != nil {
+				return result, fmt.Errorf("%w: CheckExecutionsExistActivity: %v", errors.ErrUnableToExecuteActivity, err)
+			}
+
+			if !executionsExist {
+				break
+			}
+			logger.Info("Workflow executions are still not deleted.", tag.WorkflowNamespace(params.Namespace.String()), tag.Attempt(checkAttempt))
+		}
```

Comment:
I think we can convert this part into a retried Activity (a regular one, not local activity), say EnsureNoExecutions(). This activity would return your ApplicationError if there is still executions exists. We just need to specify a retry policy with max attempt or ScheduleToClose timeout. That way, it always check with short initial backoff (like 1s), and backoff further if execution exists, so we don't have to way arbitrary delay. It will also reduce the history size because failed activity won't record. And we don't need extra logic to calculate delay time.

Reply:
This is great idea!